samples/graph-proactive-installation/csharp/proactive-Appinstallation-messages/Bots/ProactiveBot.cs

```diff
@@ -0,0 +1,223 @@
+﻿using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Configuration;
+using System;
+using Microsoft.Bot.Builder.Teams;
+using Microsoft.Bot.Schema.Teams;
+using ProactiveBot.Bots;
+using Microsoft.Bot.Connector;
+using System.Net;
+using Attachment = Microsoft.Bot.Schema.Attachment;
+using System.IO;
+using Newtonsoft.Json;
+
+namespace Microsoft.BotBuilderSamples
+{
+    public class ProactiveBot : TeamsActivityHandler
+    {
+        public static string MicrosoftTenantId;
+        public static string MicrosoftAppId;
+        public static string MicrosoftAppPassword;
+        public static string MicrosoftTeamAppid;
+        public static string  continuationToken = null;
+
+        public readonly IConfiguration _configuration;
+        ProactiveHelper Helper = new ProactiveHelper();
+        private readonly ConcurrentDictionary<string, ConversationReference> _conversationReferences;
+        public ProactiveBot(ConcurrentDictionary<string, ConversationReference> conversationReferences, IConfiguration configuration)
+        {
+            _conversationReferences = conversationReferences;
+            _configuration = configuration;
+            MicrosoftAppId = _configuration["MicrosoftAppId"];
+            MicrosoftAppPassword = _configuration["MicrosoftAppPassword"];
+            MicrosoftTeamAppid = _configuration["MicrosoftTeamAppid"];
+        }
+        private void AddConversationReference(Activity activity)
+        {
+            var conversationReference = activity.GetConversationReference();
+            _conversationReferences.AddOrUpdate(conversationReference.User.Id, conversationReference, (key, newValue) => conversationReference);
+        }
+
+        protected override async Task OnConversationUpdateActivityAsync(ITurnContext<IConversationUpdateActivity> turnContext, CancellationToken cancellationToken)
+        {
+            bool status = false;
+            string Userid = string.Empty;
+            string teamId = string.Empty;
+            string ChatId = string.Empty;
+            MicrosoftTenantId = turnContext.Activity.Conversation.TenantId;
+            var adaptiveAttachment = New_AdaptiveCardAttachment();
+            switch (turnContext.Activity.Conversation.ConversationType)
+            {
+                case "personal":
+                    Userid = turnContext.Activity.MembersAdded[0].AadObjectId;
+                    Tuple<bool, int> App_status = await Helper.AppinstallationforPersonal(Userid, MicrosoftTenantId, MicrosoftAppId, MicrosoftAppPassword, MicrosoftTeamAppid);
```

Comment:
This is getting triggered for the first time installation and failing due to 403 while calling Graph APIs. We need to request for Graph APIs first via login / application permission and then proceed with app installation.

Reply:
Need to confirm this, if we are going to add delegated auth flow to get the teamAppId generated from catalog. As far as i know application permissions is not  supported for App Catalog, if we need to add sign in flow when we can ask user to sign ?